cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java

```diff
@@ -127,24 +133,44 @@ public void startTx(Transaction tx) {
 
   @Override
   public List<Map<byte[], byte[]>> getPersisted(List<co.cask.cdap.api.dataset.table.Get> gets) {
-    List<Get> hbaseGets = Lists.transform(gets, new Function<co.cask.cdap.api.dataset.table.Get, Get>() {
-      @Nullable
-      @Override
-      public Get apply(co.cask.cdap.api.dataset.table.Get get) {
-        List<byte[]> cols = get.getColumns();
-        return createGet(get.getRow(), cols == null ? null : cols.toArray(new byte[cols.size()][]));
+    List<Get> hbaseGets = new ArrayList<>();
+    for (co.cask.cdap.api.dataset.table.Get get : gets) {
+      List<byte[]> cols = get.getColumns();
+      // Our get with empty array means get nothing. Theirs means get all.
+      // That's why we need this somewhat-messy code.
+      if (cols == null || !cols.isEmpty()) {
+        hbaseGets.add(createGet(get.getRow(), cols == null ? null : cols.toArray(new byte[cols.size()][])));
       }
-    });
+    }
+
     try {
-      Result[] results = hTable.get(hbaseGets);
-      return Lists.transform(Arrays.asList(results), new Function<Result, Map<byte[], byte[]>>() {
-        @Nullable
-        @Override
-        public Map<byte[], byte[]> apply(Result result) {
-          Map<byte[], byte[]> familyMap = result.getFamilyMap(columnFamily);
-          return familyMap != null ? familyMap : ImmutableMap.<byte[], byte[]>of();
+      List<Map<byte[], byte[]>> results =
+        Lists.transform(Arrays.asList(hTable.get(hbaseGets)), new Function<Result, Map<byte[], byte[]>>() {
+          @Nullable
+          @Override
+          public Map<byte[], byte[]> apply(Result result) {
+            Map<byte[], byte[]> familyMap = result.getFamilyMap(columnFamily);
+            return familyMap != null ? familyMap : ImmutableMap.<byte[], byte[]>of();
+          }
+        });
+
+      List<Map<byte[], byte[]>> toReturn = new ArrayList<>(gets.size());
+      int hbaseResultsIndex = 0;
+      for (int i = 0; i < gets.size(); i++) {
+        co.cask.cdap.api.dataset.table.Get get = gets.get(0);
+        List<byte[]> cols = get.getColumns();
+        if (cols != null && cols.isEmpty()) {
+          toReturn.add(ImmutableMap.<byte[], byte[]>of());
+        } else {
+          toReturn.add(results.get(hbaseResultsIndex++));
         }
-      });
+      }
+      Preconditions.checkArgument(toReturn.size() == gets.size());
+      Preconditions.checkArgument(hbaseResultsIndex == hbaseGets.size());
+      for (Map<byte[], byte[]> map : toReturn) {
+        Preconditions.checkNotNull(map);
```

Comment:
I wrote these preconditions as validation of the code, during development. I will remove them now, since they won't happen, based upon the code logic above.